source/detail/serialization/xlsx_consumer.cpp

```diff
@@ -2317,7 +2319,15 @@ void xlsx_consumer::read_stylesheet()
         if (current_style_element == qn("spreadsheetml", "borders"))
         {
             auto &borders = stylesheet.borders;
-            auto count = parser().attribute<std::size_t>("count");
+            optional<std::size_t> count;
+            if (parser().attribute_present("count"))
+            {
+                count = parser().attribute<std::size_t>("count");
+                if (count.get() <= xlnt::constants::max_elements_for_reserve())
```

Comment:
Alternative may be to reserve `xlnt::constants::max_elements_for_reserve()` elements if the requested count is larger than this maximum, in stead of just reserving nothing. Or should we really consider a value larger than max as completely trash?
What do you think?

Reply:
This is a good question that I also asked myself, but was also not sure whether it necessarily pays off. On the one hand, this would mean that the broken or maliciously crafted files would allocate too much memory unnecessarily. On the other hand, 10000 elements is not an extremely large number, so probably this would never be an issue anyway.

It would be possible to easily achieve what you said by doing something like this:
```cpp
vec.reserve(xlnt::reserve_with_limit(count.get()));
```

Where `xlnt::reserve_with_limit` would do something like this:
```cpp
size_t reserve_with_limit(size_t num_elements)
{
    return std::min(num_elements, xlnt::constants::max_elements_for_reserve());
}
```

This way, we would also make it simple for the caller (would not be much more code).

Then I however asked myself whether this is really necessary or whether I'm overthinking it, so for now I left it as it is. Thinking that most files won't be broken or malicious, it could be a nice performance increase to allocate memory right away in case a very large number of elements is necessary. Due to the fact that the exact way `std::vector::reserve` behaves is not specified in the C++ standard, I now tested the following code:
```cpp
#include <iostream>
#include <vector>

int main()
{
    constexpr size_t max_size = 10'000'000;
    size_t current_capacity = 0;
    size_t capacity_increase_count = 0;
    std::vector<char> vec;
    //vec.reserve(10000);

    for (size_t n = 0; n < max_size; ++n)
    {
        if (vec.capacity() != current_capacity)
        {
            ++capacity_increase_count;
            std::cout << "size: " << vec.size() << " | capacity: " << vec.capacity() << '\n';
            current_capacity = vec.capacity();
        }

        vec.push_back(0);
    }

    std::cout << "\nCapacity increased " << capacity_increase_count << " times.";
}
```

I tested it in the [Compiler Explorer](https://godbolt.org/), then uncommented the line with `reserve`. Results under x86-64:
| Mode | GCC 14.2  | Clang 18.1.0 | MSVC v19.40 VS17.10 |
| ------------- | ------------- | ------------- | ------------- |
| Without | 25 times | 25 times | 41 times |
| 10000 elements | 11 times | 11 times | 19 times |

So I think the wrapper around `max_elements_for_reserve()` I proposed above would make the code simple enough, while also providing a good performance increase even for a larger amount of elements. Please tell me what you think about it, and also feel free to play with the numbers :wink:

Reply:
You may choose, both are fine for me. If we would measure the execution time, the advantage of reserving 10 000 if the needed size is larger is probably less, as the reallocation is more expensive for the larger sizes. If you go for the clipping, I would only prefer a name like `clip_reserve_elements` as the name `reserve_with_limit` gives me the intention that it actually reserves something.

Reply:
By the way, due to the fact that the function no longer returns a constant, I thought about creating a new file `utils/limits.hpp` for all kinds of limits (like element reservation limits, or others). Unless you have a better idea, I'll do it like this and merge the PR afterwards :smiley:

Reply:
I don't think this file should be in the include directory (as it is an implementation detail, isn't?). I would rather place it in the same directory of `constants.php` (= detail folder) or create a utils there if desired.